text/3498-lifetime-capture-rules-2024.md

```diff
@@ -0,0 +1,642 @@
+
+# Background
+
+Rust's rules in the 2021 and earlier editions around capturing lifetimes in return position `impl Trait` (RPIT) opaque types are inconsistent, unergonomic, and not helpful to users.  In common scenarios, doing the correct thing requires a *trick* that is not well known and whose purpose is commonly not well understood.
+
+As we look forward to the 2024 edition and move toward stabilizing features such as type alias `impl Trait` (TAIT), associated type position `impl Trait` (ATPIT), return position `impl Trait` in trait (RPITIT), and `async fn` in trait (AFIT), we must decide on a clear vision of how lifetimes should be captured in Rust.
+
+We want the upcoming features in the stabilization pipeline to capture lifetimes in a way that's consistent with each other and with the way we want Rust to work and develop going forward.
+
+This RFC specifies a [solution] that achieves this.  But first, we'll describe the problem in further detail.  The descriptions and examples in this section use the semantics of Rust 2021.
+
+## Capturing lifetimes
+
+In return position `impl Trait` (RPIT) and `async fn`, an **opaque type** is a type that can only be used for its specified trait bounds (and for the "leaked" auto trait bounds of its hidden type).  A **hidden type** is the actual concrete type of the values hidden behind the opaque type.
+
+A hidden type is only allowed to name lifetime parameters when those lifetime parameters have been *"captured"* by the corresponding opaque type. For example:
+
+```rust
+// Returns: `Future<Output = &'a ()> + Captures<&'a ()>`
+async fn foo<'a>(x: &'a ()) -> &'a () { x }
+```
+
+In the above, we would say that the lifetime parameter `'a` has been captured in the returned opaque type.
+
+For an opaque type that *does not* specify an outlives bound (e.g. `+ 'other`), when a caller receives a value of that opaque type and wants to prove that it outlives some lifetime, the caller must prove that all of the captured lifetime components of the opaque type outlive that lifetime.  The captured lifetime components are the set of lifetimes contained within captured type parameters and the lifetimes represented by captured lifetime parameters.
+
+For an opaque type that *does* specify an outlives bound (e.g. `+ 'other`), when a caller receives a value of that opaque type and wants to prove that it outlives some lifetime, it's enough to prove that it outlives the specified lifetime in the bounds of the opaque after transitively taking into account all known lifetime bounds.  For such an opaque type, the *callee* must prove that all lifetime and type parameters that are used in the hidden type outlive the specified bound.
```

Comment:
> For such an opaque type, the *callee* must prove that all lifetime and type parameters that are used in the hidden type outlive the specified bound.

What this means is that this code is invalid:

```rust
// error[E0310]: the parameter type `T` may not live long enough
fn foo<T>(x: T) -> impl Sized + 'static { x }
```

`foo` here is the callee.  The hidden type is `T`.  The specified bound is `'static`.  The callee cannot prove that "all lifetime and type parameters that are used in the hidden type outlive the specified bound."  I.e., it can't prove that `T: 'static`.

Reply:
> As a minor wording nit, the structure of these two paragraphs makes it sound kind of like there are disjoint rules for opaque types that do/do not specify outlives bounds, respectively, which isn't the case. In particular, opaque types which specify outlives bounds can _also_ be reasoned about using capture rules.

While rereading this to address the nit you raised here, I found that the second paragraph should be more clear.  I've pushed a change with that clarification, and I've added an [Appendix H](https://github.com/traviscross/rfcs/blob/TC/lifetime-capture-rules-2024/text/3498-lifetime-capture-rules-2024.md#appendix-h-examples-of-outlives-rules-on-opaque-types) that explains these rules with examples.

Perhaps these changes will also more broadly answer the questions you raised above.